microprofile.ls/org.eclipse.lsp4mp.ls/src/main/java/org/eclipse/lsp4mp/model/PropertyGraph.java

```diff
@@ -92,4 +103,82 @@ public boolean hasNode(String property) {
 		return unreachable;
 	}
 
+	/**
+	 * Returns true if the graph is acyclic, and false if it has a cycle.
+	 * 
+	 * Uses DFS to figure out if the graph is acyclic.
+	 * Uses lazy loading to avoid rerunning DFS each time this is called.
+	 * 
+	 * @return true if the graph is acyclic, and false if it has a cycle.
+	 */
+	public boolean isAcyclic() {
+		if (!acyclic.isPresent()) {
+			acyclic = Optional.of(true);
+			Set<String> vertices = nodes.keySet();
+			Deque<String> searchStack = new LinkedList<>();
+			Map<String, Boolean> visited = new HashMap<>(vertices.size());
+			boolean recurseComplete;
+			for (String vertex : vertices) {
+				visited.put(vertex, false);
+			}
+			String current;
+			for (String vertex : vertices) {
+				if (!visited.get(vertex)) {
+					searchStack.addLast(vertex);
+					visited.put(vertex, true);
+					current = vertex;
+					while (searchStack.size() > 0 && acyclic.get()) {
+						current = searchStack.getLast();
+						recurseComplete = true;
+						for (String adj : nodes.get(current)) {
+							if (!visited.get(adj)) {
+								searchStack.addLast(adj);
+								visited.put(adj, true);
+								recurseComplete = false;
+								break;
+							} else if (searchStack.contains(adj)) {
+								acyclic = Optional.of(false);
+								break;
+							}
+						}
+						if (recurseComplete) {
+							searchStack.removeLast();
+						}
+					}
+				}
+				if (!acyclic.get()) {
+					break;
+				}
+			}
+		}
+		return acyclic.get();
+	}
```

Comment:
This is a very messy implementation of non-recursive DFS. I initially adapted a recursive implementation from Sedgewick and Wayne's Algorithms textbook, then discovered that the algorithm implementation was under the GPL (not compatible with the EPL). I decided to switch to a non-recursive strategy to avoid accidentally copying code, since the implementation would be fundamentally different.

Reply:
we already depend on guava, which provides a graph API. Please check https://guava.dev/releases/snapshot-jre/api/docs/com/google/common/graph/Graph.html

Reply:
Thanks! I'll rework the first property expression PR to use this too.

Reply:
Switched to Guava, cleaned up the `PropertyGraph` interface quite a bit